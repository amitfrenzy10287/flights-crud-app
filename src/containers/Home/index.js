import React from 'react';
import { connect } from 'react-redux';
import axios from '../../axios-flights';
import * as actions from '../../store/actions';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import Paper from '@material-ui/core/Paper';
import Skeleton from '@material-ui/lab/Skeleton';
import InputBase from '@material-ui/core/InputBase';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import AddIcon from '@material-ui/icons/Add';

function descendingComparator(a, b, orderBy) {
    if (b[orderBy] < a[orderBy]) {
        return -1;
    }
    if (b[orderBy] > a[orderBy]) {
        return 1;
    }
    return 0;
}

function getComparator(order, orderBy) {
    return order === 'desc'
        ? (a, b) => descendingComparator(a, b, orderBy)
        : (a, b) => -descendingComparator(a, b, orderBy);
}

function stableSort(array, comparator) {
    const stabilizedThis = array.map((el, index) => [el, index]);
    stabilizedThis.sort((a, b) => {
        const order = comparator(a[0], b[0]);
        if (order !== 0) return order;
        return a[1] - b[1];
    });
    return stabilizedThis.map((el) => el[0]);
}

const headCells = [
    { id: 'departure', numeric: true, disablePadding: false, label: 'Departure' },
    { id: 'arrival', numeric: true, disablePadding: false, label: 'Arrival' },
    { id: 'departureTime', numeric: true, disablePadding: false, label: 'Departure Time' },
    { id: 'arrivalTime', numeric: true, disablePadding: false, label: 'Arrival Time' },
    { id: 'class', numeric: true, disablePadding: false, label: 'Class' },
];

function EnhancedTableHead(props) {
    const { classes, order, orderBy, onRequestSort } = props;
    const createSortHandler = (property) => (event) => {
        onRequestSort(event, property);
    };

    return (
        <TableHead className={classes.tableHeadContainer}>
            <TableRow>
                {headCells.map((headCell) => (
                    <TableCell className={classes.tableHead}
                        key={headCell.id}
                        align={headCell.numeric ? 'right' : 'left'}
                        padding={headCell.disablePadding ? 'none' : 'default'}
                        sortDirection={orderBy === headCell.id ? order : false}
                    >
                        <TableSortLabel
                            active={orderBy === headCell.id}
                            direction={orderBy === headCell.id ? order : 'asc'}
                            onClick={createSortHandler(headCell.id)}
                        >
                            {headCell.label}
                            {orderBy === headCell.id ? (
                                <span className={classes.visuallyHidden}>
                  {order === 'desc' ? 'sorted descending' : 'sorted ascending'}
                </span>
                            ) : null}
                        </TableSortLabel>
                    </TableCell>
                ))}
            </TableRow>
        </TableHead>
    );
}

const useStyles = makeStyles((theme) => ({
    root: {
        width: '100%',
    },
    tableHeadContainer:{
        backgroundColor: '#3f51b5',
        color:'#FFFFFF'
    },
    paper: {
        width: '100%',
        marginBottom: theme.spacing(2),
    },
    table: {
        minWidth: 750,
    },
    tableHead:{
      color: '#FFFFFF',
      '& :hover':{
          color:'#FFFFFF'
      },
      '& :focus':{
         color:'#FFFFFF'
      },
      '& .MuiTableSortLabel-root.MuiTableSortLabel-active':{
        color: '#FFFFFF',
      },
      '& .MuiTableSortLabel-icon-active':{
          color: '#FFF'
      }
    },
    visuallyHidden: {
        border: 0,
        clip: 'rect(0 0 0 0)',
        height: 1,
        margin: -1,
        overflow: 'hidden',
        padding: 0,
        position: 'absolute',
        top: 20,
        width: 1,
    },
    loader:{
        width:'100%',
        height: 50
    },
    rootSearch: {
        padding: '2px 4px',
        display: 'flex',
        alignItems: 'center',
        width: 400,
        marginRight:theme.spacing(2)
    },
    input: {
        marginLeft: theme.spacing(1),
        flex: 1,
    },
    iconButton: {
        padding: 10,
    },
    divider: {
        height: 28,
        margin: 4,
    },
    searchWrap: {
        padding: theme.spacing(2)
    }

}));

const Home=(props)=>{
    const classes = useStyles();
    const {flights,loading,error} = props;
    const [order, setOrder] = React.useState('asc');
    const [orderBy, setOrderBy] = React.useState('calories');
    const [result, setResult] = React.useState({});
    const [page, setPage] = React.useState(0);
    const [rowsPerPage, setRowsPerPage] = React.useState(5);

    const handleRequestSort = (event, property) => {
        const isAsc = orderBy === property && order === 'asc';
        setOrder(isAsc ? 'desc' : 'asc');
        setOrderBy(property);
    };

    const handleChangePage = (event, newPage) => {
        setPage(newPage);
    };

    const handleChangeRowsPerPage = (event) => {
        setRowsPerPage(parseInt(event.target.value, 10));
        setPage(0);
    };

    React.useEffect(()=>{
        props.getAllFlights();
    },[]);

    const reformatDateTime = React.useCallback(((flightDate)=>{
        if(!flightDate) return false;
        if((typeof flightDate)==="string") return flightDate;
        let unix_timestamp = flightDate;
        let date = new Date(unix_timestamp * 1000);
        let hours = date.getHours();
        let minutes = "0" + date.getMinutes();
        let seconds = "0" + date.getSeconds();
        return  hours + ':' + minutes.substr(-2) + ':' + seconds.substr(-2);
    }),[]);

    const finalTableData = React.useMemo(()=>{
        const count = Object.keys(flights).length;
        if(count===0 || error ===true) return [];
        return flights;
    },[flights,loading,error,page]);

    const searchData = React.useCallback((e)=>
    {
        const filteredResult =
            finalTableData
            .filter(s => {
                let arrival = s.arrival.toLowerCase();
                let departure = s.departure.toLowerCase();
                let arrTime =reformatDateTime(s.departureTime);
                let depTime = reformatDateTime(s.arrivalTime);
                let val = e.target.value.toLowerCase();
                return arrival.includes(val) || departure.includes(val);
            });
        setResult(prevState=>({
            ...prevState,
            ...filteredResult
        }));
    },[finalTableData]);
    const tableData = Object.keys(result).length > 0 ? Object.keys(result).map((k) => result[k]) : finalTableData;
    const addFlight=React.useCallback(()=>{
        props.history.push('/addflight');
    },[]);
    return (
        <div className={classes.root}>
            <Paper className={classes.paper}>
                <Grid className={classes.searchWrap} container item justify="right">
                <Paper className={classes.rootSearch}>
                    <IconButton className={classes.iconButton} aria-label="menu">
                        <MenuIcon />
                    </IconButton>
                    <InputBase
                        className={classes.input}
                        placeholder="Search"
                        inputProps={{ 'aria-label': 'search' }}
                        onChange={e=>searchData(e)}
                    />
                    <IconButton className={classes.iconButton} aria-label="search">
                        <SearchIcon />
                    </IconButton>
                </Paper>
                <Button variant="contained" onClick={addFlight} color="primary">
                    <AddIcon /> Add Flight
                </Button>
                </Grid>
                <TableContainer>
                    <Table
                        className={classes.table}
                        aria-labelledby="tableTitle"
                        size={'medium'}
                        aria-label="enhanced table"
                    >
                        <EnhancedTableHead
                            classes={classes}
                            order={order}
                            orderBy={orderBy}
                            onRequestSort={handleRequestSort}
                            rowCount={tableData.length}
                        />
                        <TableBody>
                            {tableData.length > 0 ? stableSort(tableData, getComparator(order, orderBy))
                                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                                .map((row, index) => {
                                    return (
                                        <TableRow
                                            hover
                                            key={index}
                                        >
                                            <TableCell align="right">{row.departure}</TableCell>
                                            <TableCell align="right">{row.arrival}</TableCell>
                                            <TableCell align="right">{reformatDateTime(row.departureTime)}</TableCell>
                                            <TableCell align="right">{reformatDateTime(row.arrivalTime)}</TableCell>
                                            <TableCell align="right">
                                                {row.type && row.type==="cheap"? 'Economy Class': 'Business Class'}
                                            </TableCell>
                                        </TableRow>
                                    );
                                }):(loading===true ?
                                    <TableCell colSpan={5}>
                                        <Skeleton className={classes.loader} animation="wave" />
                                        <Skeleton className={classes.loader} animation="wave" />
                                        <Skeleton className={classes.loader} animation="wave" />
                                        <Skeleton className={classes.loader} animation="wave" />
                                        <Skeleton className={classes.loader} animation="wave" />
                                    </TableCell>
                                :<TableCell colSpan={6}>
                                  No Records found!
                                </TableCell>)
                            }
                        </TableBody>
                    </Table>
                </TableContainer>
                <TablePagination
                    rowsPerPageOptions={[5, 10, 25]}
                    component="div"
                    count={tableData.length}
                    rowsPerPage={rowsPerPage}
                    page={page}
                    onChangePage={handleChangePage}
                    onChangeRowsPerPage={handleChangeRowsPerPage}
                />
            </Paper>
        </div>
    );
};

const mapStateToProps = state => {
    return {
        flights: state.flights.allFlights,
        error: state.flights.error,
        loading: state.flights.loading
    }
};

const mapDispatchToProps = dispatch => {
    return {
        getAllFlights: () => dispatch(actions.getAllFlights()),
    }
};

export default connect(mapStateToProps, mapDispatchToProps)( Home, axios );