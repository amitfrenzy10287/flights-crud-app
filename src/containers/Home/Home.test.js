import React from 'react';
import Home from './index';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import SearchIcon from '@material-ui/icons/Search';
import IconButton from '@material-ui/core/IconButton';

configure({adapter: new Adapter()});
const mockStore = configureMockStore();
const store = mockStore({});

describe("To test the title exist", () => {
    it("should render without throwing an error", () => {
        expect(
            shallow(
                <Provider store={store}>
                    <Home />
                </Provider>
            ).exists(
                <IconButton aria-label="search">
                    <SearchIcon />
                </IconButton>
            )
        );
    });
});

describe('<Home /> with no props', () => {
    const state = {
        flights:{
            allFlights: {}
        }
    };
    const container = shallow(
        <Provider store={store}>
            <Home state={state} />
        </Provider>
    );
    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('should have a btn component', ()=> {
        //There should be only one button
        expect(container.find('Button'));
    });

});