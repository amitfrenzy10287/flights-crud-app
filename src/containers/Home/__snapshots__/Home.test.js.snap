// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> with no props should match the snapshot 1`] = `"<div class=\\"makeStyles-root-1\\"><div class=\\"MuiPaper-root makeStyles-paper-3 MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiGrid-root makeStyles-searchWrap-12 MuiGrid-container MuiGrid-item\\"><div class=\\"MuiPaper-root makeStyles-rootSearch-8 MuiPaper-elevation1 MuiPaper-rounded\\"><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-10\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"menu\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z\\"></path></svg></span></button><div class=\\"MuiInputBase-root makeStyles-input-9\\"><input type=\\"text\\" placeholder=\\"Search\\" aria-label=\\"search\\" class=\\"MuiInputBase-input\\"/></div><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-10\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"search\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path></svg></span></button></div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z\\"></path></svg> Add Flight</span></button></div><div class=\\"MuiTableContainer-root\\"><table class=\\"MuiTable-root makeStyles-table-4\\" aria-labelledby=\\"tableTitle\\" aria-label=\\"enhanced table\\"><thead class=\\"MuiTableHead-root makeStyles-tableHeadContainer-2\\"><tr class=\\"MuiTableRow-root MuiTableRow-head\\"><th class=\\"MuiTableCell-root MuiTableCell-head makeStyles-tableHead-5 MuiTableCell-alignRight\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Departure<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th><th class=\\"MuiTableCell-root MuiTableCell-head makeStyles-tableHead-5 MuiTableCell-alignRight\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Arrival<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th><th class=\\"MuiTableCell-root MuiTableCell-head makeStyles-tableHead-5 MuiTableCell-alignRight\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Departure Time<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th><th class=\\"MuiTableCell-root MuiTableCell-head makeStyles-tableHead-5 MuiTableCell-alignRight\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Arrival Time<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th><th class=\\"MuiTableCell-root MuiTableCell-head makeStyles-tableHead-5 MuiTableCell-alignRight\\" scope=\\"col\\"><span class=\\"MuiButtonBase-root MuiTableSortLabel-root\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">Class<svg class=\\"MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path></svg></span></th></tr></thead><tbody class=\\"MuiTableBody-root\\"><td class=\\"MuiTableCell-root MuiTableCell-body\\" colSpan=\\"6\\">No Records found!</td></tbody></table></div><div class=\\"MuiTablePagination-root\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters\\"><div class=\\"MuiTablePagination-spacer\\"></div><p class=\\"MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit\\">Rows per page:</p><div class=\\"MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot\\"><div class=\\"MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input\\" tabindex=\\"0\\" role=\\"button\\" aria-haspopup=\\"listbox\\" aria-label=\\"Rows per page:\\">5</div><input type=\\"hidden\\" value=\\"5\\"/><svg class=\\"MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></div><p class=\\"MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit\\">1-NaN of undefined</p><div class=\\"MuiTablePagination-actions\\"><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\" title=\\"Previous page\\" aria-label=\\"Previous page\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z\\"></path></svg></span></button><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" title=\\"Next page\\" aria-label=\\"Next page\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z\\"></path></svg></span></button></div></div></div></div></div>"`;
